File: sdk-web/src/embedding-cache.ts
import type { CachedEmbedding, ProximityResult } from "./types.js";

/** Squared Euclidean distance: ||a - b||² */
export function squaredEuclidean(a: number[], b: number[]): number {
  let sum = 0;
  for (let i = 0; i < a.length; i++) {
    const d = a[i] - b[i];
    sum += d * d;
  }
  return sum;
}

/**
 * ETag-based embedding cache.
 * Syncs advertiser embeddings from the server and provides local proximity search.
 */
export class EmbeddingCache {
  private embeddings: CachedEmbedding[] = [];
  private etag: string | null = null;

  constructor(private endpoint: string) {}

  /** Current cached embeddings. */
  getEmbeddings(): CachedEmbedding[] {
    return this.embeddings;
  }

  /**
   * Sync embeddings from the server.
   * Uses If-None-Match / ETag for efficient caching — 304 means no update needed.
   */
  async sync(): Promise<void> {
    const headers: Record<string, string> = {};
    if (this.etag) {
      headers["If-None-Match"] = this.etag;
    }

    const resp = await fetch(`${this.endpoint}/embeddings`, { headers });
    if (resp.status === 304) return;

    if (!resp.ok) {
      throw new Error(`Embeddings API error: ${resp.status}`);
    }

    const data = await resp.json();
    this.embeddings = data.embeddings;
    this.etag = resp.headers.get("ETag");
  }

  /**
   * Compute squared Euclidean distance from a query embedding to all cached
   * advertiser embeddings. Returns results sorted ascending (closest first).
   */
  proximity(queryEmbedding: number[]): ProximityResult[] {
    return this.embeddings
      .map((entry) => ({
        id: entry.id,
        distance: squaredEuclidean(queryEmbedding, entry.embedding),
      }))
      .sort((a, b) => a.distance - b.distance);
  }
}

File: sdk-android/src/main/kotlin/dev/cloudx/sdk/EmbeddingCache.kt
package dev.vectorspace.sdk

import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject

/**
 * Manages local caching of advertiser embeddings with ETag-based
 * conditional fetching (If-None-Match / 304 Not Modified).
 */
internal class EmbeddingCache(
    private val client: OkHttpClient,
    private val endpoint: String
) {
    private var embeddings: List<EmbeddingEntry> = emptyList()
    private var etag: String? = null

    /**
     * Fetch embeddings from the server. Uses ETag for 304 caching so
     * repeated calls are cheap when nothing has changed.
     */
    fun sync() {
        val requestBuilder = Request.Builder()
            .url("$endpoint/embeddings")
            .get()

        etag?.let { requestBuilder.header("If-None-Match", it) }

        val response = client.newCall(requestBuilder.build()).execute()
        response.use { resp ->
            if (resp.code == 304) {
                return // cache is fresh
            }

            if (!resp.isSuccessful) {
                throw VectorSpaceException("Failed to fetch embeddings: HTTP ${resp.code}")
            }

            val body = resp.body?.string()
                ?: throw VectorSpaceException("Empty response body from /embeddings")
            val json = JSONObject(body)
            val embeddingsArray = json.getJSONArray("embeddings")

            val entries = mutableListOf<EmbeddingEntry>()
            for (i in 0 until embeddingsArray.length()) {
                val entry = embeddingsArray.getJSONObject(i)
                val id = entry.getString("id")
                val embeddingJson = entry.getJSONArray("embedding")
                val embedding = FloatArray(embeddingJson.length()) { j ->
                    embeddingJson.getDouble(j).toFloat()
                }
                entries.add(EmbeddingEntry(id, embedding))
            }

            embeddings = entries
            resp.header("ETag")?.let { etag = it }
        }
    }

    /**
     * Compute squared Euclidean distance from [queryEmbedding] to each
     * cached embedding. Returns results sorted ascending by distance.
     */
    fun proximity(queryEmbedding: FloatArray): List<ProximityResult> {
        return embeddings.map { entry ->
            ProximityResult(
                id = entry.id,
                distance = squaredEuclidean(queryEmbedding, entry.embedding)
            )
        }.sortedBy { it.distance }
    }

    /**
     * ||a - b||^2
     */
    private fun squaredEuclidean(a: FloatArray, b: FloatArray): Float {
        var sum = 0f
        val len = minOf(a.size, b.size)
        for (i in 0 until len) {
            val diff = a[i] - b[i]
            sum += diff * diff
        }
        return sum
    }
}

/**
 * Exception thrown by the VectorSpace SDK for server or protocol errors.
 */
class VectorSpaceException(message: String, cause: Throwable? = null) : Exception(message, cause)
